src/script.c

```diff
@@ -1235,6 +1235,32 @@ int wally_witness_p2wpkh_from_sig(
     return ret;
 }
 
+int wally_witness_p2tr_from_sig(
+    const unsigned char *sig,
+    size_t sig_len,
+    struct wally_tx_witness_stack **witness)
+{
+    int ret;
+
+    if (witness)
+        *witness = NULL;
+
+    /* Required to be a valid BIP340 length of 64 + possible sighash flag */
+    if (!sig || (sig_len != 64 && sig_len != 65) || !witness)
+        return WALLY_EINVAL;
+
+    ret = wally_tx_witness_stack_init_alloc(1, witness);
+
+    if (ret == WALLY_OK) {
+        return wally_tx_witness_stack_add(*witness, sig, sig_len);
+    } else {
+        wally_tx_witness_stack_free(*witness);
+        *witness = NULL;
+    }
```

Comment:
Almost! :)
If alloc fails there's nothing to free, all wally calls clean up after themselves (although cleaning up again is harmless). You need to catch the case where alloc is ok, but add fails, i.e. something like:

```
    if ((ret = wally_tx_witness_stack_init_alloc(1, witness)) == WALLY_OK) {
        ret = wally_tx_witness_stack_add(*witness, sig, sig_len);
        if (ret != WALLY_OK) {
            wally_tx_witness_stack_free(*witness);
            *witness = NULL;
        }
    }
```